Added argsort and sort method to SparsePauliOp

kUmezawa opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [☑] I have added the tests to cover my changes.
- [ ☑] I have updated the documentation accordingly.
- [ ☑] I have read the CONTRIBUTING document.
-->

### Summary
This is an update as part of QAMP's activities.
Relating issue: https://github.com/qiskit-advocate/qamp-spring-22/issues/20
PauliList has argsort() method and sort() method.
We need these features in SparsePauliOp.

### Details and comments
argsort() method returns the composition of permutations in the order of sorting by coefficient and sorting by Pauli.

sort() method returns a new sorted SparsePauliOp class. After sorting the coefficients using numpy's argsort, sort by Pauli.

kUmezawa commented:
@ikkoham 
Please review. Thank you.

HuangJunye commented:
@kUmezawa Is this part of QAMP? If so can you please mention the QAMP project in the description? I will also ad `QAMP` label to the PR. Thank you.

kUmezawa commented:
@HuangJunye 
>Is this part of QAMP? If so can you please mention the QAMP project in the description?
Yes, this is part of QAMP. I have updated this pr description. Thank you.

ikkoham reviewed (changes requested):
Thank you. Originally, the simplify method sorted as out of spec, but since [0.20.0](https://github.com/Qiskit/qiskit-terra/pull/7656) it no longer sorts, so this method is useful.

Basically, It looks good to me, but release note is required for new feature. Could you please add the release note?

kUmezawa commented:
@ikkoham 
Is it okay to add a new yml file to releasenotes / notes / 0.20?

t-imamichi commented:
There is an attempt to parametrize coefficients of `SparsePauliOp` #7215.
Does `weight` option of this PR work fine with parameter expressions as well as complex values?

t-imamichi commented:
As for release note, please check out this documentation.
https://github.com/Qiskit/qiskit-terra/blob/main/CONTRIBUTING.md#release-notes

kUmezawa commented:
@t-imamichi
>Does weight option of this PR work fine with parameter expressions as well as complex values?

I haven't considered it yet.
@ikkoham Let's discuss it in QAMP.

 >As for release note, please check out this documentation.

Thank you.
I follow the steps to create the release notes

kUmezawa commented:
@t-imamichi 
I consulted with Hamamura-sann
Since the Parameter has nothing to do with this weight, we judge that it is not necessary to consider it this time.

ikkoham commented:
Thanks. I'll have to think about whether we handle the error or simply just get a TypeError from NumPy when introducing Parameter for SparsePauliOp. I'll keep this in my mind.

Cryoris commented on test/python/quantum_info/operators/symplectic/test_sparse_pauli_op.py:
```diff
@@ -507,6 +507,129 @@ def test_simplify_zero(self, num_qubits):
         np.testing.assert_array_equal(zero_op.paulis.phase, np.zeros(zero_op.size))
         np.testing.assert_array_equal(simplified_op.paulis.phase, np.zeros(simplified_op.size))
 
+    def test_sort(self):
```
Could you add a test for a corner case like an empty SparsePauliOp to make sure no errors are being raised?

kUmezawa replied:
OK. I will add a test.

Cryoris commented on qiskit/quantum_info/operators/symplectic/sparse_pauli_op.py:
```diff
@@ -438,6 +438,74 @@ def simplify(self, atol=None, rtol=None):
+    def argsort(self, weight=False):
+        """Return indices for sorting the rows of the table.
+
+        Returns the composition of permutations in the order of sorting
```
I think it would be very helpful to add a code snippet that demonstrates the effect of setting `weight` to true or false, could you add one? 🙂

kUmezawa replied:
OK. I will add code snippe. Thank you.

Cryoris reviewed (commented):
Also from my side the code LGTM! But what is the motivation for adding these methods? That wasn't quite clear from the PR description 🙂 

I'm quite confused by the `weight` argument -- I thought it was related to the coefficients and only when looking at `PauliList` I realized that you define the weight as the number of non-identity Paulis. In the context of the `PauliList` I see how that makes sense but since we have coefficients in the `SparsePauliOp` I think this naming might be a bit unclear. Could we maybe use a more precise argument?

kUmezawa commented:
@Cryoris 

Thank you for your review.

>Also from my side the code LGTM! But what is the motivation for adding these methods? That wasn't quite clear from the PR description

I'm sorry that the motive is difficult to understand. As Mr. Hamamura wrote, the simplify method cannot be sorted, so the motivation was to prepare an alternative method. I will specify it in the release notes. 


>Could we maybe use a more precise argument?

OK. I will consider it.

kUmezawa commented on releasenotes/notes/0.20/add-sparsepauliop-methods-00a7e6cc7055e1d0.yaml:
```diff
@@ -0,0 +1,16 @@
+---
+features:
+  - |
+    Added a new argsort method.
+    Returns the composition of permutations in the order of sorting 
+    by coefficient and sorting by Pauli. By using the `weight` kwarg 
+    the output can additionally be sorted 
+    by the number of non-identity terms in the Pauli, where the set of
+    all Pauli's of a given weight are still ordered lexicographically.
+  - |
+    Added a new sort method.
```
I'm sorry that the motive is difficult to understand. As Mr. Hamamura wrote, the simplify method cannot be sorted, so the motivation was to prepare an alternative method. I will specify it in the release notes.

kUmezawa commented:
@Cryoris @ikkoham 
Sorry for the late response.
The release notes have been fixed.
Added test code for empty list.
Added sample code to the documentation.

Thank you.

Cryoris commented on qiskit/quantum_info/operators/symplectic/sparse_pauli_op.py:
```diff
@@ -444,6 +444,115 @@ def simplify(self, atol=None, rtol=None):
+
+        **Example**
+
+        Here is an example of how to use SparcePauliOp argsort.
```
```suggestion
        Here is an example of how to use SparsePauliOp argsort.
```

Cryoris commented on qiskit/quantum_info/operators/symplectic/sparse_pauli_op.py:
```diff
@@ -444,6 +444,115 @@ def simplify(self, atol=None, rtol=None):
+
+        Args:
+            weight (bool): optionally sort by weight if True (Default: False).
+            This argument is the same as weight of argsort method of PauliList
```
```suggestion
                This argument is the same as weight of argsort method of :class:`.PauliList`.
```

Cryoris commented on qiskit/quantum_info/operators/symplectic/sparse_pauli_op.py:
```diff
@@ -444,6 +444,115 @@ def simplify(self, atol=None, rtol=None):
+
+        Args:
+            weight (bool): optionally sort by weight if True (Default: False).
+            This argument is the same as weight of sort method of PauliList
```
See comment on the documentation of `argsort` 🙂

Cryoris commented on test/python/quantum_info/operators/symplectic/test_sparse_pauli_op.py:
```diff
@@ -510,6 +510,132 @@ def test_simplify_zero(self, num_qubits):
 
+    def test_sort(self):
+        """Test sort method."""
+        with self.assertRaises(QiskitError):
+            target = SparsePauliOp([], [])
```
Shouldn't sorting on an empty set work? I would argue since it's empty it's already "sorted" anyways 😄

Cryoris commented:
> I understand that the meaning of your advice is to point out what happens when you sort SparcePauliOp with an empty list. Is that wrong?

I mean that right now it looks like sorting an empty SparsePauliOp raises an error. But I think it should not raise an error and just return the empty SparsePauliOp 🙂

kUmezawa commented:
@Cryoris 
Thank you.

>I mean that right now it looks like sorting an empty SparsePauliOp raises an error. But I think it should not raise an error and just return the empty SparsePauliOp

I get an error when generating a SparcePauliOp class with an empty list.
I don't think it's possible to sort an empty SparcePauliOp.

![image](https://user-images.githubusercontent.com/12453501/177956358-dc0a030c-fe95-4ba6-a4d1-eccb56723099.png)

Cryoris commented:
Oh fair enough! Then of course sorting doesn't work. Thanks for trying! 🙂

kUmezawa commented:
@Cryoris
Thank you.

＞Oh fair enough! Then of course sorting doesn't work. Thanks for trying!

OK. Thank you. 

＞See comment on the documentation of argsort

Added description of arguments.

ikkoham commented:
right. empty `SparsePauliOp` is not allowed.

Cryoris reviewed (approved):
LGTM thanks for the updates @kUmezawa!

ikkoham reviewed (approved):
LGTM. Thank you for your contribution!
